File: backend/app/schemas/comparison.py
"""Comparison schemas — strategy comparison request/response."""

from __future__ import annotations

from pydantic import BaseModel, Field


class StrategyConfig(BaseModel):
    type: str = Field(
        ...,
        pattern=r"^(top|portfolio|random|wheeling|budget|profile|method)$",
    )
    count: int = Field(10, gt=0, le=100)
    # Optional params depending on type
    numbers: list[int] | None = None
    stars: list[int] | None = None
    guarantee: int | None = None
    profile: str | None = None
    method: str | None = None
    budget: float | None = None
    objective: str | None = None


class ComparisonRequest(BaseModel):
    strategies: list[StrategyConfig] = Field(..., min_length=2, max_length=5)
    include_gain_scenarios: bool = False


class StrategyMetrics(BaseModel):
    type: str
    label: str
    grid_count: int
    grids: list[dict] | None = None
    avg_score: float | None = None
    score_variance: float | None = None
    diversity: float | None = None
    coverage: float | None = None
    cost: float = 0.0
    robustness: float | None = None
    expected_gain: float | None = None


class ComparisonSummary(BaseModel):
    best_score: str | None = None
    best_diversity: str | None = None
    best_coverage: str | None = None
    best_cost: str | None = None
    recommendation: str


class ComparisonResponse(BaseModel):
    strategies: list[StrategyMetrics]
    summary: ComparisonSummary

File: frontend/src/types/comparison.ts
/** Comparison types for strategy comparison. */

export interface StrategyConfig {
  type:
    | "top"
    | "portfolio"
    | "random"
    | "wheeling"
    | "budget"
    | "profile"
    | "method";
  count?: number;
  numbers?: number[];
  stars?: number[];
  guarantee?: number;
  profile?: string;
  method?: string;
  budget?: number;
  objective?: string;
}

export interface ComparisonRequest {
  strategies: StrategyConfig[];
  include_gain_scenarios?: boolean;
}

export interface StrategyMetrics {
  type: string;
  label: string;
  grid_count: number;
  grids: Array<{ numbers: number[]; stars?: number[] | null }> | null;
  avg_score: number | null;
  score_variance: number | null;
  diversity: number | null;
  coverage: number | null;
  cost: number;
  robustness: number | null;
  expected_gain: number | null;
}

export interface ComparisonSummary {
  best_score: string | null;
  best_diversity: string | null;
  best_coverage: string | null;
  best_cost: string | null;
  recommendation: string;
}

export interface ComparisonResponse {
  strategies: StrategyMetrics[];
  summary: ComparisonSummary;
}
